astropy/table/operations.py

```diff
@@ -786,22 +732,25 @@ def _vstack(arrays, join_type='outer', col_name_map=None):
     lens = [len(arr) for arr in arrays]
     n_rows = sum(lens)
     out = _get_out_class(arrays)(masked=masked)
-    out_descrs = get_descrs(arrays, col_name_map)
-    for out_descr in out_descrs:
-        name = out_descr[0]
-        dtype = out_descr[1:]
-        if masked:
-            out[name] = ma.array(data=np.zeros(n_rows, dtype),
-                                 mask=np.ones(n_rows, ma.make_mask_descr(dtype)))
-        else:
-            out[name] = np.empty(n_rows, dtype=dtype)
 
     for out_name, in_names in six.iteritems(col_name_map):
+        # List of input arrays that contribute to this output column
+        cols = [arr[name] for arr, name in zip(arrays, in_names) if name is not None]
+
+        col_cls = _get_out_class(cols)
+        try:
+            out[out_name] = col_cls.empty_like(cols, n_rows, metadata_conflicts, out_name)
```

Comment:
To do: I think it is more clear to check for an `empty_like` method here instead of pre-checking above.